### ISSUE TITLE
[MPS] einsum 42x slower since 1.13.0.dev20220925 — on (16, 4096, 40)*(16, 40, 4096) matmul

### ISSUE THREAD

**Issue by Birch-san** [pytorch/pytorch#87010]
### 🐛 Describe the bug

_spun off from https://github.com/pytorch/pytorch/issues/85224#issuecomment-1261529333._

einsum perf has regressed, as of `1.13.0.dev20220925` nightly. regression is present in [`1.13.0` release candidate](https://download.pytorch.org/whl/test/cpu/torch-1.13.0-cp310-none-macosx_11_0_arm64.whl).  
suspected to be caused by https://github.com/pytorch/pytorch/pull/84890 (most relevant-looking commit of that nightly, not proven beyond doubt).  
not related to `opt_einsum` (it's uninstalled).

```python
import torch
import time
from torch import einsum

# MPS backend has no synchronization API comparable to CUDA's:
# https://pytorch.org/docs/master/notes/cuda.html#asynchronous-execution
# so we have to get creative.
# to force Python to wait for computation to complete: introduce a data-dependency on every element in the tensor.
baseline_start = time.perf_counter()
torch.rand(16, 4096, 4096, dtype=torch.float, device="mps").max().item()
print('for reference: our sync operation -- on a tensor of our target size -- is responsible for only %.4f seconds of overhead' % (time.perf_counter()-baseline_start))

repeats = 10

ein0_batch_duration = 0
for ix in range(repeats):
  q = torch.rand(16, 4096, 40, dtype=torch.float, device="mps")
  k = torch.rand(16, 4096, 40, dtype=torch.float, device="mps")
  ein0_start = time.perf_counter()
  einsum('b i d, b j d -> b i j', q, k).max().item()
  ein0_duration = time.perf_counter()-ein0_start
  print('einsum 0 iteration %d took %.4f seconds' % (ix, ein0_duration))
  ein0_batch_duration += ein0_duration
print('%d iterations of einsum 0 took %.4f seconds; avg %.4f secs' % (repeats, ein0_batch_duration, ein0_batch_duration/repeats))

ein1_batch_duration = 0
for ix in range(repeats):
  attn = torch.rand(16, 4096, 4096, dtype=torch.float, device="mps")
  v = torch.rand(16, 4096, 40, dtype=torch.float, device="mps")
  ein1_start = time.perf_counter()
  einsum('b i j, b j d -> b i d', attn, v).max().item()
  ein1_duration = time.perf_counter()-ein1_start
  print('einsum 1 iteration %d took %.4f seconds' % (ix, ein1_duration))
  ein1_batch_duration += ein1_duration
print('%d iterations of einsum 1 took %.4f seconds; avg %.4f secs' % (repeats, ein1_batch_duration, ein1_batch_duration/repeats))
```

1.13.0:

```
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.1172 seconds of overhead
einsum 0 iteration 0 took 1.0375 seconds
einsum 0 iteration 1 took 0.6801 seconds
einsum 0 iteration 2 took 0.6819 seconds
einsum 0 iteration 3 took 0.6785 seconds
einsum 0 iteration 4 took 0.6775 seconds
einsum 0 iteration 5 took 0.6762 seconds
einsum 0 iteration 6 took 0.6772 seconds
einsum 0 iteration 7 took 0.6744 seconds
einsum 0 iteration 8 took 0.6788 seconds
einsum 0 iteration 9 took 0.6774 seconds
10 iterations of einsum 0 took 7.1397 seconds; avg 0.7140 secs
einsum 1 iteration 0 took 0.0504 seconds
einsum 1 iteration 1 took 0.0499 seconds
einsum 1 iteration 2 took 0.0498 seconds
einsum 1 iteration 3 took 0.0491 seconds
einsum 1 iteration 4 took 0.0498 seconds
einsum 1 iteration 5 took 0.0500 seconds
einsum 1 iteration 6 took 0.0500 seconds
einsum 1 iteration 7 took 0.0486 seconds
einsum 1 iteration 8 took 0.0619 seconds
einsum 1 iteration 9 took 0.0488 seconds
10 iterations of einsum 1 took 0.5081 seconds; avg 0.0508 secs
```

Torch 1.12.1:

```
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.1467 seconds of overhead
einsum 0 iteration 0 took 0.0856 seconds
einsum 0 iteration 1 took 0.0095 seconds
einsum 0 iteration 2 took 0.0096 seconds
einsum 0 iteration 3 took 0.0092 seconds
einsum 0 iteration 4 took 0.0093 seconds
einsum 0 iteration 5 took 0.0093 seconds
einsum 0 iteration 6 took 0.0093 seconds
einsum 0 iteration 7 took 0.0094 seconds
einsum 0 iteration 8 took 0.0092 seconds
einsum 0 iteration 9 took 0.0093 seconds
10 iterations of einsum 0 took 0.1696 seconds; avg 0.0170 secs
einsum 1 iteration 0 took 0.0397 seconds
einsum 1 iteration 1 took 0.0892 seconds
einsum 1 iteration 2 took 0.0648 seconds
einsum 1 iteration 3 took 0.0409 seconds
einsum 1 iteration 4 took 0.0387 seconds
einsum 1 iteration 5 took 0.0403 seconds
einsum 1 iteration 6 took 0.0405 seconds
einsum 1 iteration 7 took 0.0403 seconds
einsum 1 iteration 8 took 0.0398 seconds
einsum 1 iteration 9 took 0.0409 seconds
10 iterations of einsum 1 took 0.4750 seconds; avg 0.0475 secs
```

einsum 0 is 42x slower in the 1.13.0 release candidate.  
einsum 1 is 1.07x slower in the 1.13.0 release candidate.  
_7% slower… could that explain https://github.com/pytorch/pytorch/issues/85297?_

einsum 0 notable for involving a transpose. I tried reformulating it to eliminate this factor. still slow:  
https://github.com/pytorch/pytorch/issues/85224#issuecomment-1279542405
so perhaps more significant is the combination of tensor sizes in the matmul: `(16, 4096, 40)*(16, 40, 4096)`.

context in which this is used is the [cross-attention algorithm in stable-diffusion](https://github.com/CompVis/stable-diffusion/blob/69ae4b35e0a0f6ee1af8bb9a5d0016ccb27e36dc/ldm/modules/attention.py#L170-L193).  
overall impact is that image generation is 6x slower (now takes 1min rather than 10secs to generate a draft-quality image).

these einsums can also be formulated as matmuls:  
https://github.com/Birch-san/stable-diffusion/commit/d2d533dbc3fe2e430a3ab9feaf23aa28a8b8178f

but matmul perf is slow too. stable-diffusion generation times with einsum vs with matmul:  
1.12.1 einsum (10.7s) > 1.12.1, 1.13.0 matmul (14.6s) > 1.13.0 einsum (71.6s)

### Versions

```
PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.30)
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.10.4 (main, Mar 31 2022, 03:37:37) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.5-arm64-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] open-clip-torch==2.0.2
[pip3] pytorch-lightning==1.7.6
[pip3] torch==1.13.0
[pip3] torch-fidelity==0.3.0
[pip3] torchdiffeq==0.2.3
[pip3] torchmetrics==0.7.0
[pip3] torchvision==0.15.0.dev20221013
[conda] numpy                     1.23.1          py310h220015d_0  
[conda] numpy-base                1.23.1          py310h742c864_0  
[conda] open-clip-torch           2.0.2                    pypi_0    pypi
[conda] pytorch-lightning         1.7.6                    pypi_0    pypi
[conda] torch                     1.13.0                   pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torchdiffeq               0.2.3                    pypi_0    pypi
[conda] torchmetrics              0.7.0                    pypi_0    pypi
[conda] torchvision               0.15.0.dev20221013          pypi_0    pypi
```

cc @VitalyFedyunin @ngimel @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by janeyx99** [comment]
I tried replicating this on Linux CUDA + CPU with the torch.benchmark library, but the performance difference for our einsum 0 example is slight across nightlies, so **we can confirm this regression is MPS specific/at least not on Linux CPU + A100 CUDA**.
```
equation: b i d, b j d -> b i j         operands: [torch.Size([16, 4096, 40]), torch.Size([16, 4096, 40])]
equation: b i j, b j d -> b i d         operands: [torch.Size([16, 4096, 4096]), torch.Size([16, 4096, 40])]
[---------------------------------------------------------  ---------------------------------------------------------]
                             |  torch.einsum 1.12.1  |  torch.einsum 9.24  |  torch.einsum 9.28  |  torch.einsum 10.14
1 threads: -----------------------------------------------------------------------------------------------------------
      b i d, b j d -> b i j  |          3.7          |        2.972        |        2.970        |        2.970       
      b i j, b j d -> b i d  |          3.8          |        3.826        |        3.826        |        3.827       

Times are in milliseconds (ms).
```

Similarly, for Linux CPU:
```
equation: b i d, b j d -> b i j         operands: [torch.Size([16, 4096, 40]), torch.Size([16, 4096, 40])]
equation: b i j, b j d -> b i d         operands: [torch.Size([16, 4096, 4096]), torch.Size([16, 4096, 40])]
[---------------------------------------------------------  ---------------------------------------------------------]
                             |  torch.einsum 1.12.1  |  torch.einsum 9.24  |  torch.einsum 9.28  |  torch.einsum 10.14
1 threads: -----------------------------------------------------------------------------------------------------------
      b i d, b j d -> b i j  |          1609         |         1610        |         1634        |         1609       
      b i j, b j d -> b i d  |           213         |          210        |          209        |          211       

Times are in milliseconds (ms).
```

If you're curious, my code (for the CUDA):
```
#! /bin/bash/python3
# -*- coding: utf-8 -*-
import torch
import torch.utils.benchmark as benchmark

def generate_test_case_from_85224(dtype):
    q = torch.rand(16, 4096, 40, device='cuda', dtype=dtype)
    k = torch.rand(16, 4096, 40, device='cuda', dtype=dtype)
    yield ('b i d, b j d -> b i j', '', q, k)
    attn = torch.rand(16, 4096, 4096, device='cuda', dtype=dtype)
    v = torch.rand(16, 4096, 40, device='cuda', dtype=dtype)
    yield ('b i j, b j d -> b i d', '', attn, v)

def main():
    test_case_generators = [
        generate_test_case_from_85224(torch.float)
    ]
    results = []

    for test_case_generator in test_case_generators:
        for equation, descriptor, *operands in test_case_generator:
            print('equation:', equation, '\toperands:', [op.shape for op in operands])

            results.append(benchmark.Timer(
                stmt='torch.einsum(equation, *operands)',
                setup='import subprocess\nimport sys\nsubprocess.check_call([sys.executable, "-m", "pip", "install", "--pre", "torch==1.12.1", "--extra-index-url", "https://download.pytorch.org/whl/nightly/cu116"])',
                globals={'equation': equation, 'operands': operands},
                sub_label=equation + descriptor,
                description='torch.einsum 1.12.1',
            ).blocked_autorange())

            results.append(benchmark.Timer(
                stmt='torch.einsum(equation, *operands)',
                setup='import subprocess\nimport sys\nsubprocess.check_call([sys.executable, "-m", "pip", "install", "--pre", "torch==1.13.0.dev20220924", "--extra-index-url", "https://download.pytorch.org/whl/nightly/cu116"])',
                globals={'equation': equation, 'operands': operands},
                sub_label=equation + descriptor,
                description='torch.einsum 9.24',
            ).blocked_autorange())

            results.append(benchmark.Timer(
                stmt='torch.einsum(equation, *operands)',
                setup='import subprocess\nimport sys\nsubprocess.check_call([sys.executable, "-m", "pip", "install", "--pre", "torch==1.13.0.dev20220928", "--extra-index-url", "https://download.pytorch.org/whl/nightly/cu116"])',
                globals={'equation': equation, 'operands': operands},
                sub_label=equation + descriptor,
                description='torch.einsum 9.28',
            ).blocked_autorange())

            results.append(benchmark.Timer(
                stmt='torch.einsum(equation, *operands)',
                setup='import subprocess\nimport sys\nsubprocess.check_call([sys.executable, "-m", "pip", "install", "--pre", "torch==1.14.0.dev20221014", "--extra-index-url", "https://download.pytorch.org/whl/nightly/cu116"])',
                globals={'equation': equation, 'operands': operands},
                sub_label=equation + descriptor,
                description='torch.einsum 10.14',
            ).blocked_autorange())

    compare = benchmark.Compare(results)
    compare.trim_significant_figures()
    compare.colorize(rowwise=True)
    compare.print()

if __name__ == "__main__":
    main()
```

**Comment by Vargol** [comment]
So, I've compiled PyTorch from master and run @Birch-san 's test code, you'll have to excuse the slow timings 8Gb M1 here :-)

```
% python test.py 
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.2277 seconds of overhead
einsum 0 iteration 0 took 7.7644 seconds
einsum 0 iteration 1 took 13.1047 seconds
einsum 0 iteration 2 took 14.7310 seconds
einsum 0 iteration 3 took 14.0011 seconds
einsum 0 iteration 4 took 16.5059 seconds
einsum 0 iteration 5 took 16.8674 seconds
einsum 0 iteration 6 took 15.9724 seconds
einsum 0 iteration 7 took 14.0594 seconds
einsum 0 iteration 8 took 14.1236 seconds
einsum 0 iteration 9 took 13.4323 seconds
10 iterations of einsum 0 took 140.5622 seconds; avg 14.0562 secs
einsum 1 iteration 0 took 2.4364 seconds
einsum 1 iteration 1 took 0.1886 seconds
einsum 1 iteration 2 took 0.1792 seconds
einsum 1 iteration 3 took 0.2007 seconds
einsum 1 iteration 4 took 0.1825 seconds
einsum 1 iteration 5 took 0.1796 seconds
einsum 1 iteration 6 took 0.1858 seconds
einsum 1 iteration 7 took 0.1793 seconds
einsum 1 iteration 8 took 0.1869 seconds
einsum 1 iteration 9 took 0.1830 seconds
10 iterations of einsum 1 took 4.1019 seconds; avg 0.4102 secs

```
I then took the ensin function and 
bool einsum_check_label(unsigned char label) 
uint8_t einsum_label_to_index(unsigned char label) 
unsigned char einsum_index_to_label(uint8_t index) 
from tag 1.12.1

put then back it, tested it complied and ran, getting the same figures (as expected as it still running the current code).

I then took the signature from the new version of einsum, used it for the old version and added an extra parameters to the new version, so that the old version of einsum is now called, complied and ran.

```
% python test.py                    
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.1953 seconds of overhead
einsum 0 iteration 0 took 0.1850 seconds
einsum 0 iteration 1 took 0.0773 seconds
einsum 0 iteration 2 took 0.0774 seconds
einsum 0 iteration 3 took 0.0776 seconds
einsum 0 iteration 4 took 0.0773 seconds
einsum 0 iteration 5 took 0.0775 seconds
einsum 0 iteration 6 took 0.0782 seconds
einsum 0 iteration 7 took 0.0773 seconds
einsum 0 iteration 8 took 0.0770 seconds
einsum 0 iteration 9 took 0.0778 seconds
10 iterations of einsum 0 took 0.8822 seconds; avg 0.0882 secs
einsum 1 iteration 0 took 0.1531 seconds
einsum 1 iteration 1 took 0.1652 seconds
einsum 1 iteration 2 took 0.1529 seconds
einsum 1 iteration 3 took 0.1694 seconds
einsum 1 iteration 4 took 0.1580 seconds
einsum 1 iteration 5 took 0.1699 seconds
einsum 1 iteration 6 took 0.1725 seconds
einsum 1 iteration 7 took 0.1710 seconds
einsum 1 iteration 8 took 0.1533 seconds
einsum 1 iteration 9 took 0.1562 seconds
10 iterations of einsum 1 took 1.6216 seconds; avg 0.1622 secs
```

Fairly convincing evidence has the kwarg's change caused some of the performance regression.

Is there a way to change the code so that if path argument is not used it calls the old version ?

**Comment by brkirch** [comment]
commit e78e00f4d98c4376e298902db8aae7e7057e86df (parent of e7e1cd945fe218fde228cedbdb1509f1750f70ea)
```
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.1244 seconds of overhead
einsum 0 iteration 0 took 0.1291 seconds
einsum 0 iteration 1 took 0.0183 seconds
einsum 0 iteration 2 took 0.0159 seconds
einsum 0 iteration 3 took 0.0156 seconds
einsum 0 iteration 4 took 0.0158 seconds
einsum 0 iteration 5 took 0.0160 seconds
einsum 0 iteration 6 took 0.0159 seconds
einsum 0 iteration 7 took 0.0159 seconds
einsum 0 iteration 8 took 0.0158 seconds
einsum 0 iteration 9 took 0.0160 seconds
10 iterations of einsum 0 took 0.2742 seconds; avg 0.0274 secs
einsum 1 iteration 0 took 0.0476 seconds
einsum 1 iteration 1 took 0.0920 seconds
einsum 1 iteration 2 took 0.0477 seconds
einsum 1 iteration 3 took 0.0484 seconds
einsum 1 iteration 4 took 0.0482 seconds
einsum 1 iteration 5 took 0.0482 seconds
einsum 1 iteration 6 took 0.0480 seconds
einsum 1 iteration 7 took 0.0479 seconds
einsum 1 iteration 8 took 0.0478 seconds
einsum 1 iteration 9 took 0.0481 seconds
10 iterations of einsum 1 took 0.5239 seconds; avg 0.0524 secs
```
commit e7e1cd945fe218fde228cedbdb1509f1750f70ea
```
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.1233 seconds of overhead
einsum 0 iteration 0 took 1.0263 seconds
einsum 0 iteration 1 took 0.7207 seconds
einsum 0 iteration 2 took 0.6720 seconds
einsum 0 iteration 3 took 0.6807 seconds
einsum 0 iteration 4 took 0.6771 seconds
einsum 0 iteration 5 took 0.6720 seconds
einsum 0 iteration 6 took 0.6719 seconds
einsum 0 iteration 7 took 0.6732 seconds
einsum 0 iteration 8 took 0.6763 seconds
einsum 0 iteration 9 took 0.6767 seconds
10 iterations of einsum 0 took 7.1469 seconds; avg 0.7147 secs
einsum 1 iteration 0 took 0.0550 seconds
einsum 1 iteration 1 took 0.0542 seconds
einsum 1 iteration 2 took 0.0540 seconds
einsum 1 iteration 3 took 0.0548 seconds
einsum 1 iteration 4 took 0.0547 seconds
einsum 1 iteration 5 took 0.0548 seconds
einsum 1 iteration 6 took 0.0549 seconds
einsum 1 iteration 7 took 0.0669 seconds
einsum 1 iteration 8 took 0.0548 seconds
einsum 1 iteration 9 took 0.0546 seconds
10 iterations of einsum 1 took 0.5585 seconds; avg 0.0559 secs
```
> Is there a way to change the code so that if path argument is not used it calls the old version ?

No idea if that is something that would be considered acceptable as a temporary fix by the PyTorch team, but that could be done with something like this (this is the einsum function in 5099883f059a9b15592b8ba3b7bf83145163b966, latest commit at the time of writing, with a conditional added that uses the code from the einsum function from e78e00f4d98c4376e298902db8aae7e7057e86df if no path argument is given):
```
Tensor einsum(c10::string_view equation, TensorList operands, at::OptionalIntArrayRef path) {
  if (path.has_value()) {
    TORCH_CHECK(!operands.empty(), "einsum(): must provide at least one operand");
    const auto num_ops = operands.size();

    //if (path.has_value()) {
    const auto path_size = num_ops == 1 ? 1 : (num_ops - 1) * 2;
    TORCH_CHECK(
        path->size() == path_size,
        "einsum(): expected contraction path given in path parameter to have size ",
        path_size,
        " but got ",
        path->size());
    //}

    // Labels must be in range [A-Za-z]
    constexpr uint8_t NUM_OF_LETTERS = 'z' - 'a' + 1;
    constexpr uint8_t TOTAL_LABELS = NUM_OF_LETTERS * 2;

    // Code used to identify ELLIPSIS ("...")
    constexpr uint8_t ELLIPSIS = TOTAL_LABELS;

    // Convert label in [A-Za-z] to subscript in [0, TOTAL_LABELS)
    auto label_to_subscript = [=](unsigned char label) -> uint8_t {
      return std::isupper(label) ? label - 'A' : label - 'a' + NUM_OF_LETTERS;
    };

    // Convert subscript in [0, TOTAL_LABELS) to label in [A-Za-z]
    auto subscript_to_label = [=](uint8_t s) -> unsigned char {
      return s < NUM_OF_LETTERS ? s + 'A' : s + 'a' - NUM_OF_LETTERS;
    };

    // Find arrow (->) to split equation into lhs and rhs
    const auto arrow_pos = equation.find("->");
    const auto lhs = equation.substr(0, arrow_pos);

    // Convert labels for input operands into an index in [0, 52) and store
    // them in op_labels for each operand along with ELLIPSIS if present.
    std::vector<std::vector<uint8_t>> op_labels(num_ops);
    bool ell_in_input = false;
    std::size_t curr_op = 0;
    for (std::size_t i = 0; i < lhs.length(); ++i) {
      const unsigned char label = lhs[i];
      switch (label) {
        case ' ':
          // Ignore spaces
          break;

        case '.':
          TORCH_CHECK(
              // Only one ellipsis per operand can be given
              !ell_in_input,
              "einsum(): found \'.\' for operand ",
              curr_op,
              " for which an ellipsis was already found");
          TORCH_CHECK(
              // Ensure it's a valid ellipsis
              i + 2 < lhs.length() && lhs[++i] == '.' && lhs[++i] == '.',
              "einsum(): found \'.\' for operand ",
              curr_op,
              " that is not part of any ellipsis");
          op_labels[curr_op].push_back(ELLIPSIS);
          ell_in_input = true;
          break;

        case ',':
          // Move onto next operand
          ++curr_op;
          TORCH_CHECK(
              curr_op < num_ops,
              "einsum(): fewer operands were provided than specified in the equation");
          ell_in_input = false;
          break;

        default:
          // Parse label
          TORCH_CHECK(
              std::isalpha(label),
              "einsum(): invalid subscript given at index ",
              i,
              " in the equation string, subscripts must be in [a-zA-Z]");
          op_labels[curr_op].push_back(label_to_subscript(label));
      }
    }

    TORCH_CHECK(
        curr_op == num_ops - 1,
        "einsum(): more operands were provided than specified in the equation");

    std::vector<int64_t> label_count(TOTAL_LABELS, 0);

    // The maximum number of dimensions covered by any ellipsis, needed when
    // unsqueezing missing dimensions from operands to permute and broadcast
    int64_t ell_num_dim = 0;

    // Compute label frequency and number of dimensions covered by ellipsis
    // We do this after parsing labels to make it more readable and simpler
    // to compute the number of dimensions covered by ellipsis.
    for(const auto i : c10::irange(num_ops)) {
      const auto operand = operands[i];
      const auto labels = op_labels[i];
      const auto ndims = operand.dim();
      int64_t nlabels = static_cast<int64_t>(labels.size());
      bool has_ellipsis = false;

      for (const auto& label : labels) {
        if (label == ELLIPSIS) {
          --nlabels;
          has_ellipsis = true;
          ell_num_dim = std::max(ell_num_dim, ndims - nlabels);
        } else {
          ++label_count[label];
        }
      }

      TORCH_CHECK(
          has_ellipsis ? nlabels <= ndims : nlabels == ndims,
          "einsum(): the number of subscripts in the equation (",
          nlabels,
          has_ellipsis ? ") is more than the number of dimensions ("
                       : ") does not match the number of dimensions (",
          ndims,
          ") for operand ",
          i,
          has_ellipsis ? "" : " and no ellipsis was given");
    }

    // We want to align the dimensions of every input tensor to have
    // shape out_dims + sum_dims. For this, we create a mapping of label
    // to index into the permuted shape.
    std::vector<int64_t> label_perm_index(TOTAL_LABELS, -1);

    // Current index in the permuted shape
    int64_t perm_index = 0;

    // Start index of ellipsis dimensions in the permuted shape
    int64_t ell_index = 0;
    bool ell_in_output = false;

    if (arrow_pos == std::string::npos) {
      // Implicit output is ellipsis (...) + labels seen only once
      perm_index = ell_num_dim;
      // ell_in_output is used to stop us from reducing ellipses dims later
      ell_in_output = true;
      for (const auto label : c10::irange(TOTAL_LABELS)) {
        if (label_count[label] == 1) {
          label_perm_index[label] = perm_index++;
        }
      }
    } else {
      // Parse explicit output
      const auto rhs = equation.substr(arrow_pos + 2);
      for (std::size_t i = 0; i < rhs.length(); ++i) {
        const unsigned char label = rhs[i];
        switch (label) {
          case ' ':
            // Ignore spaces
            break;

          case '.':
            TORCH_CHECK(
                // There can only be one ellipsis in the output
                !ell_in_output,
                "einsum(): found \'.\' for output but an ellipsis (...) was already found");
            TORCH_CHECK(
                // Ensure ellipsis is correct
                i + 2 < rhs.length() && rhs[++i] == '.' && rhs[++i] == '.',
                "einsum(): found \'.\' for output that is not part of any ellipsis (...)");
            ell_index = perm_index;
            perm_index += ell_num_dim;
            ell_in_output = true;
            break;

          default:
            TORCH_CHECK(
                std::isalpha(label),
                "einsum(): invalid subscript given at index ",
                lhs.size() + 2 + i,
                " in the equation string, subscripts must be in [a-zA-Z]");
            const auto index = label_to_subscript(label);
            TORCH_CHECK(
                // Ensure label appeared at least once for some input operand and at
                // most once for the output
                label_count[index] > 0 && label_perm_index[index] == -1,
                "einsum(): output subscript ",
                label,
                label_perm_index[index] > -1
                    ? " appears more than once in the output"
                    : " does not appear in the equation for any input operand");
            label_perm_index[index] = perm_index++;
        }
      }
    }

    // Save number of dimensions in output before adding contraction dims (dims to sum out)
    const int64_t out_num_dim = perm_index;

    // If ellipsis is not part of the output, add to contraction dimensions
    if (!ell_in_output) {
      ell_index = perm_index;
      perm_index += ell_num_dim;
    }

    // Add contraction labels (labels not present in output)
    for (const auto label : c10::irange(TOTAL_LABELS)) {
      if (label_count[label] > 0 && label_perm_index[label] == -1) {
        label_perm_index[label] = perm_index++;
      }
    }

    // Next: we check the sizes, take diagonals for repeated labels, unsqueeze
    // missing dimensions so all operands have the same dimensions and permute
    // the operands to align the dimensions following the indices computed above.
    // We also count how many operands have dimension with size != 1 for each
    // label used to identify which dimensions can be contracted.
    std::vector<SymInt> label_size(TOTAL_LABELS, 1);
    std::vector<SymInt> ell_sizes(ell_num_dim, 1);
    std::vector<uint64_t> dim_counts(perm_index, 0);
    std::vector<Tensor> ops;
    for (const auto i : irange(num_ops)) {
      auto op = operands[i];
      std::vector<int64_t> permutation(perm_index, -1);
      std::int64_t dim = 0;
      for (const auto s : op_labels[i]) {
        if (s == ELLIPSIS) {
          // Iterate over each dimension covered by ellipsis
          const auto ndim = operands[i].ndimension() - (static_cast<int64_t>(op_labels[i].size()) - 1);
          for (auto j = ell_num_dim - ndim; j < ell_num_dim; ++j) {
            if (op.sym_size(dim) != 1) {
              // Update ellipsis size
              TORCH_CHECK(
                  ell_sizes[j] == 1 || ell_sizes[j] == op.sym_size(dim),
                  "einsum(): dimension ",
                  dim,
                  " covered by ellipsis in operand ",
                  i,
                  "has size ",
                  op.size(dim),
                  " which does not broadcast with previously seen ellipsis with size ",
                  ell_sizes[j],
                  " for the respective dimension");
              ell_sizes[j] = op.sym_size(dim);
              ++dim_counts[ell_index + j];
            }
            permutation[ell_index + j] = dim++;
          }
        } else if (permutation[label_perm_index[s]] == -1) {
          if (op.sym_size(dim) != 1) {
            // Update subscript
            TORCH_CHECK(
                label_size[s] == 1 || label_size[s] == op.sym_size(dim),
                "einsum(): subscript ",
                subscript_to_label(s),
                " has size ",
                op.sym_size(dim),
                " for operand ",
                i,
                " which does not broadcast with previously seen size ",
                label_size[s]);
            label_size[s] = op.sym_size(dim);
            ++dim_counts[label_perm_index[s]];
          }
          permutation[label_perm_index[s]] = dim++;
        } else {
          // Repeated label, take diagonal
          const auto prev_dim = permutation[label_perm_index[s]];
          TORCH_CHECK(
            op.sym_size(dim) == op.sym_size(prev_dim),
              "einsum(): subscript ",
              subscript_to_label(s),
              " is repeated for operand ",
              i,
              " but the sizes don't match, ",
              op.sym_size(dim),
              " != ",
              op.sym_size(prev_dim));
          op = op.diagonal(0, prev_dim, dim).movedim(-1, prev_dim);
        }
      }

      // Add dimensions for missing labels
      for (auto& val : permutation) {
        if (val == -1) {
          op = op.unsqueeze(dim);
          val = dim++;
        }
      }
      ops.emplace_back(op.permute(permutation));
    }

    const auto contract_path = path.value_or(std::vector<int64_t>{});
    auto it = contract_path.begin();

    // Contract
    while (ops.size() > 1) {
      int64_t i = 0;
      int64_t j = 1;

      //if (path.has_value()) {
      i = *it++;
      j = *it++;
      if (j < i) {
        std::swap(i, j);
      }

      TORCH_CHECK(
          i != j && i >= 0 && j < static_cast<int64_t>(ops.size()),
          "einsum(): invalid contraction (",
          i,
          ", ",
          j,
          i == j ? ") cannot contract an operand with itself"
                 : ") operand index is out of bounds");
      //}

      auto a = ops[i];
      auto b = ops[j];
      ops.erase(ops.begin() + j);
      ops.erase(ops.begin() + i);

      // Collect dimensions that can be summed now
      std::vector<int64_t> sum_dims;
      SmallVector<int64_t, 5> a_dims_to_sum;
      SmallVector<int64_t, 5> b_dims_to_sum;
      for (auto dim = out_num_dim; dim < perm_index; ++dim) {
        if (a.sym_size(dim) != 1 && b.sym_size(dim) != 1) {
          if (--dim_counts[dim] == 1) {
            sum_dims.push_back(dim);
            dim_counts[dim] = 0;
          }
        } else if (dim_counts[dim] == 1) {
          if (a.sym_size(dim) != 1) {
            a_dims_to_sum.push_back(dim);
            dim_counts[dim] = 0;
          } else if (b.sym_size(dim) != 1) {
            b_dims_to_sum.push_back(dim);
            dim_counts[dim] = 0;
          }
        }
      }

      // Sum multiple dims at a time to minimize the number of kernel calls to sum
      if (!a_dims_to_sum.empty()) {
        a = a.sum(a_dims_to_sum, true);
      }
      if (!b_dims_to_sum.empty()) {
        b = b.sum(b_dims_to_sum, true);
      }

      ops.emplace_back(sumproduct_pair(a, b, sum_dims, true));
    }

    // Sum out contraction dims
    if (perm_index - out_num_dim > 0) {
      std::vector<int64_t> sum_dims(perm_index - out_num_dim);
      std::iota(sum_dims.begin(), sum_dims.end(), out_num_dim);
      ops[0] = ops[0].sum(sum_dims);
    }

    return ops[0];
  }
  else
  {
    TORCH_CHECK(!operands.empty(), "einsum(): must provide at least one operand");

    // Labels must be in range [A-Za-z]
    constexpr uint8_t NUM_OF_LETTERS = 'z' - 'a' + 1;
    constexpr uint8_t TOTAL_LABELS = NUM_OF_LETTERS * 2;

    // Code used to identify ELLIPSIS ("...")
    constexpr uint8_t ELLIPSIS = TOTAL_LABELS;

    // Convert label in [A-Za-z] to subscript in [0, TOTAL_LABELS)
    auto label_to_subscript = [=](unsigned char label) -> uint8_t {
      return std::isupper(label) ? label - 'A' : label - 'a' + NUM_OF_LETTERS;
    };

    // Convert subscript in [0, TOTAL_LABELS) to label in [A-Za-z]
    auto subscript_to_label = [=](uint8_t s) -> unsigned char {
      return s < NUM_OF_LETTERS ? s + 'A' : s + 'a' - NUM_OF_LETTERS;
    };

    // Find arrow (->) to split equation into lhs and rhs
    const auto arrow_pos = equation.find("->");
    const auto lhs = equation.substr(0, arrow_pos);

    const auto num_ops = operands.size();

    // Convert labels for input operands into an index in [0, 52) and store
    // them in op_labels for each operand along with ELLIPSIS if present.
    std::vector<std::vector<uint8_t>> op_labels(num_ops);
    bool found_ell = false;
    std::size_t curr_op = 0;
    for (auto i = decltype(lhs.length()){0}; i < lhs.length(); ++i) {
      const unsigned char label = lhs[i];
      switch (label) {
        case ' ':
          // Ignore spaces
          break;

        case '.':
          TORCH_CHECK(
              // Only one ellipsis per operand can be given
              !found_ell,
              "einsum(): found \'.\' for operand ",
              curr_op,
              " for which an ellipsis was already found");
          TORCH_CHECK(
              // Ensure it's a valid ellipsis
              i + 2 < lhs.length() && lhs[++i] == '.' && lhs[++i] == '.',
              "einsum(): found \'.\' for operand ",
              curr_op,
              " that is not part of any ellipsis");
          op_labels[curr_op].push_back(ELLIPSIS);
          found_ell = true;
          break;

        case ',':
          // Move onto next operand
          ++curr_op;
          TORCH_CHECK(
              curr_op < num_ops,
              "einsum(): fewer operands were provided than specified in the equation");
          found_ell = false;
          break;

        default:
          // Parse label
          TORCH_CHECK(
              std::isalpha(label),
              "einsum(): invalid subscript given at index ",
              i,
              " in the equation string, subscripts must be in [a-zA-Z]");
          op_labels[curr_op].push_back(label_to_subscript(label));
      }
    }

    TORCH_CHECK(
        curr_op == num_ops - 1,
        "einsum(): more operands were provided than specified in the equation");

    std::vector<int64_t> label_count(TOTAL_LABELS, 0);

    // The maximum number of dimensions covered by any ellipsis, needed when
    // unsqueezing missing dimensions from operands to permute and broadcast
    int64_t ell_num_dim = 0;

    // Compute label frequency and number of dimensions covered by ellipsis
    // We do this after parsing labels to make it more readable and simpler
    // to compute the number of dimensions covered by ellipsis.
    for(const auto i : c10::irange(num_ops)) {
      const auto operand = operands[i];
      const auto labels = op_labels[i];
      const auto ndims = operand.dim();
      int64_t nlabels = static_cast<int64_t>(labels.size());
      bool has_ellipsis = false;

      for (const auto& label : labels) {
        if (label == ELLIPSIS) {
          --nlabels;
          has_ellipsis = true;
          ell_num_dim = std::max(ell_num_dim, ndims - nlabels);
        } else {
          ++label_count[label];
        }
      }

      TORCH_CHECK(
          has_ellipsis ? nlabels <= ndims : nlabels == ndims,
          "einsum(): the number of subscripts in the equation (",
          nlabels,
          has_ellipsis ? ") is more than the number of dimensions ("
                       : ") does not match the number of dimensions (",
          ndims,
          ") for operand ",
          i,
          has_ellipsis ? "" : " and no ellipsis was given");
    }

    // We want to align the dimensions of every input tensor to have
    // shape out_dims + sum_dims. For this, we create a mapping of label
    // to index into the permuted shape.
    std::vector<int64_t> label_perm_index(TOTAL_LABELS, -1);

    // Current index in the permuted shape
    int64_t perm_index = 0;

    // Start index of ellipsis dimensions in the permuted shape
    int64_t ell_index = 0;
    found_ell = false;

    if (arrow_pos == std::string::npos) {
      // Implicit output is ellipsis (...) + labels seen only once
      perm_index = ell_num_dim;
      found_ell = true;
      for (const auto label : c10::irange(TOTAL_LABELS)) {
        if (label_count[label] == 1) {
          label_perm_index[label] = perm_index++;
        }
      }
    } else {
      // Parse explicit output
      const auto rhs = equation.substr(arrow_pos + 2);
      for (auto i = decltype(rhs.length()){0}; i < rhs.length(); ++i) {
        const unsigned char label = rhs[i];
        switch (label) {
          case ' ':
            // Ignore spaces
            break;

          case '.':
            TORCH_CHECK(
                // There can only be one ellipsis in the output
                !found_ell,
                "einsum(): found \'.\' for output but an ellipsis (...) was already found");
            TORCH_CHECK(
                // Ensure ellipsis is correct
                i + 2 < rhs.length() && rhs[++i] == '.' && rhs[++i] == '.',
                "einsum(): found \'.\' for output that is not part of any ellipsis (...)");
            ell_index = perm_index;
            perm_index += ell_num_dim;
            found_ell = true;
            break;

          default:
            TORCH_CHECK(
                std::isalpha(label),
                "einsum(): invalid subscript given at index ",
                lhs.size() + 2 + i,
                " in the equation string, subscripts must be in [a-zA-Z]");
            const auto index = label_to_subscript(label);
            TORCH_CHECK(
                // Ensure label appeared at least once for some input operand and at
                // most once for the output
                label_count[index] > 0 && label_perm_index[index] == -1,
                "einsum(): output subscript ",
                label,
                label_perm_index[index] > -1
                    ? " appears more than once in the output"
                    : " does not appear in the equation for any input operand");
            label_perm_index[index] = perm_index++;
        }
      }
    }

    // Save output size before adding contraction dims (dims to sum out)
    const int64_t out_size = perm_index;

    // If ellipsis is not part of the output, add to contraction dimensions
    if (!found_ell) {
      ell_index = perm_index;
      perm_index += ell_num_dim;
    }

    // Add contraction labels (labels not present in output)
    for (const auto label : c10::irange(TOTAL_LABELS)) {
      if (label_count[label] > 0 && label_perm_index[label] == -1) {
        label_perm_index[label] = perm_index++;
      }
    }

    // Here we unsqueeze missing dimensions to make all operands have the same
    // number of dimensions. We take diagonals for repeated labels within the
    // same operand. Finally we permute the operands to align dimensions as
    // per the perm_out_index we computed above.
    std::vector<Tensor> permuted_operands;
    for (const auto i: c10::irange(num_ops)) {
      std::vector<int64_t> perm_shape(perm_index, -1);
      std::vector<int64_t> label_dim(TOTAL_LABELS, -1);
      Tensor operand = operands[i];
      const auto labels = op_labels[i];
      const auto original_sizes = operand.sizes();

      int64_t j = 0;
      for (const auto& label : labels) {
        if (label == ELLIPSIS) {
          // Add missing dimensions covered by the ellipsis
          const auto num_missing_dim =
              ell_num_dim - (original_sizes.size() - labels.size() + 1);
          for (const auto k : c10::irange(num_missing_dim)) {
            (void)k; //Suppress unused warning
            operand = operand.unsqueeze(j);
          }
          for (const auto k : c10::irange(ell_num_dim)) {
            perm_shape[ell_index + k] = j++;
          }
        } else if (label_dim[label] != -1) {
          // Repeated label, take diagonal
          const auto dim = label_dim[label];
          TORCH_CHECK(
              operand.size(j) == operand.size(dim),
              "einsum(): subscript ",
              subscript_to_label(label),
              " is repeated for operand ",
              i,
              " but the sizes don't match, ",
              operand.size(j),
              " != ",
              operand.size(dim));
          operand = operand.diagonal(0, dim, j).movedim(-1, dim);
        } else {
          // Lookup output index for label
          label_dim[label] = j;
          perm_shape[label_perm_index[label]] = j++;
        }
      }

      // Add dimensions for missing labels
      for (int64_t& index : perm_shape) {
        if (index == -1) {
          operand = operand.unsqueeze(-1);
          index = j++;
        }
      }

      permuted_operands.push_back(operand.permute(perm_shape));
    }

    // Check if operands broadcast and keep track of last operand with
    // dimension size != 1 for optimizing reductions
    std::vector<std::size_t> dim_last_op(perm_index, 0);
    bool has_zero_size_dim = false;
    for (const auto dim : c10::irange(perm_index)) {
      auto broadcast_size = permuted_operands[0].size(dim);
      for (const auto i: c10::irange(1, num_ops)) {
        const auto dim_size = permuted_operands[i].size(dim);
        if (broadcast_size != dim_size && broadcast_size != 1 && dim_size != 1) {
          std::ostringstream msg;
          msg << "einsum(): operands do not broadcast with remapped shapes [original->remapped]:";
          for (const auto j: c10::irange(num_ops)) {
            msg << " " << operands[j].sizes() << "->"
                << permuted_operands[j].sizes();
          }
          TORCH_CHECK(false, msg.str());
        }
        if (dim_size != 1) {
          broadcast_size = dim_size;
          dim_last_op[dim] = i;
        }
      }
      has_zero_size_dim |= broadcast_size == 0;
    }

    // Compute result
    Tensor result = permuted_operands[0];

    // Fast path for when an operand has zero sized dim
    if (has_zero_size_dim) {
      std::vector<int64_t> out_shape(out_size);
      for (const auto i : c10::irange(out_size)) {
        out_shape[i] = permuted_operands[dim_last_op[i]].size(i);
      }
      return at::zeros(out_shape, result.options());
    }

    // Sum out or squeeze dimensions that are size 1 for all later operands
    int64_t dim = out_size;
    for (int64_t i = dim; i < perm_index; ++i, ++dim) {
      if (dim_last_op[i] == 0) {
        if (result.size(dim) == 1) {
          result = result.squeeze(dim--);
        } else {
          result = result.sum(dim--);
        }
      }
    }

    for (const auto i: c10::irange(1, num_ops)) {
      Tensor operand = permuted_operands[i];
      std::vector<int64_t> sum_dims;

      // Sum out or squeeze dimensions that are size 1 for all later operands
      dim = out_size;
      for (int64_t j = dim; j < perm_index; ++j, ++dim) {
        if (dim_last_op[j] < i) {
          operand = operand.squeeze(dim);
          --dim;
        } else if (dim_last_op[j] == i) {
          if (result.size(dim) == 1) {
            operand = operand.sum(dim);
            result = result.squeeze(dim);
            --dim;
          } else {
            sum_dims.push_back(dim);
          }
        }
      }

      // Multiply tensors and sum out dimensions in sum_dims
      if (sum_dims.empty()) {
        result = result.mul(operand);
      } else if (sum_dims.size() == result.sizes().size()) {
        result = result.flatten().dot(operand.flatten());
      } else {
        result = sumproduct_pair(result, operand, sum_dims, false);
      }
    }

    return result;
  }
}
```

Here the benchmark using that code with 5099883f059a9b15592b8ba3b7bf83145163b966
```
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.0810 seconds of overhead
einsum 0 iteration 0 took 0.0927 seconds
einsum 0 iteration 1 took 0.0159 seconds
einsum 0 iteration 2 took 0.0154 seconds
einsum 0 iteration 3 took 0.0155 seconds
einsum 0 iteration 4 took 0.0154 seconds
einsum 0 iteration 5 took 0.0154 seconds
einsum 0 iteration 6 took 0.0155 seconds
einsum 0 iteration 7 took 0.0154 seconds
einsum 0 iteration 8 took 0.0153 seconds
einsum 0 iteration 9 took 0.0155 seconds
10 iterations of einsum 0 took 0.2320 seconds; avg 0.0232 secs
einsum 1 iteration 0 took 0.0416 seconds
einsum 1 iteration 1 took 0.0421 seconds
einsum 1 iteration 2 took 0.0416 seconds
einsum 1 iteration 3 took 0.0417 seconds
einsum 1 iteration 4 took 0.0420 seconds
einsum 1 iteration 5 took 0.0423 seconds
einsum 1 iteration 6 took 0.0418 seconds
einsum 1 iteration 7 took 0.0420 seconds
einsum 1 iteration 8 took 0.0422 seconds
einsum 1 iteration 9 took 0.0423 seconds
10 iterations of einsum 1 took 0.4196 seconds; avg 0.0420 secs
```

**Comment by Birch-san** [comment]
@janeyx99  
finally got round to measuring latest einsum perf (`torch-1.14.0.dev20221023`):

```
for reference: our sync operation -- on a tensor of our target size -- is responsible for only 0.0977 seconds of overhead
einsum 0 iteration 0 took 0.1428 seconds
einsum 0 iteration 1 took 0.0167 seconds
einsum 0 iteration 2 took 0.0174 seconds
einsum 0 iteration 3 took 0.0172 seconds
einsum 0 iteration 4 took 0.0172 seconds
einsum 0 iteration 5 took 0.0171 seconds
einsum 0 iteration 6 took 0.0172 seconds
einsum 0 iteration 7 took 0.0170 seconds
einsum 0 iteration 8 took 0.0170 seconds
einsum 0 iteration 9 took 0.0166 seconds
10 iterations of einsum 0 took 0.2962 seconds; avg 0.0296 secs
einsum 1 iteration 0 took 0.0439 seconds
einsum 1 iteration 1 took 0.0433 seconds
einsum 1 iteration 2 took 0.0432 seconds
einsum 1 iteration 3 took 0.0435 seconds
einsum 1 iteration 4 took 0.0436 seconds
einsum 1 iteration 5 took 0.0438 seconds
einsum 1 iteration 6 took 0.0439 seconds
einsum 1 iteration 7 took 0.0431 seconds
einsum 1 iteration 8 took 0.0435 seconds
einsum 1 iteration 9 took 0.0442 seconds
10 iterations of einsum 1 took 0.4360 seconds; avg 0.0436 secs
```

compared to `1.13.0.dev20220924` (the final nightly before kwarg path optimization):

einsum 0 is 2.7% faster.  
einsum 1 is 22% faster.

so, perf regression more than fixed. 🎉 

but we're still slower than pytorch 1.12.1 stable, probably due to still-unresolved https://github.com/pytorch/pytorch/issues/85297.

compared to 1.12.1 stable:

einsum 0 is 74% slower.  
einsum 1 is 8.9% faster. 🎉 

overall, stable-diffusion image generation on `torch-1.14.0.dev20221023` takes 12.3 secs, which is 15% slower than on 1.12.1 stable.

**Comment by Birch-san** [comment]
```
einsum 0 iteration 0 took 0.1356 seconds
einsum 0 iteration 1 took 0.0164 seconds
einsum 0 iteration 2 took 0.0168 seconds
einsum 0 iteration 3 took 0.0169 seconds
einsum 0 iteration 4 took 0.0167 seconds
einsum 0 iteration 5 took 0.0168 seconds
einsum 0 iteration 6 took 0.0167 seconds
einsum 0 iteration 7 took 0.0166 seconds
einsum 0 iteration 8 took 0.0167 seconds
einsum 0 iteration 9 took 0.0167 seconds
10 iterations of einsum 0 took 0.2858 seconds; avg 0.0286 secs
einsum 1 iteration 0 took 0.0435 seconds
einsum 1 iteration 1 took 0.0431 seconds
einsum 1 iteration 2 took 0.0429 seconds
einsum 1 iteration 3 took 0.0438 seconds
einsum 1 iteration 4 took 0.0433 seconds
einsum 1 iteration 5 took 0.0432 seconds
einsum 1 iteration 6 took 0.0436 seconds
einsum 1 iteration 7 took 0.0436 seconds
einsum 1 iteration 8 took 0.0428 seconds
einsum 1 iteration 9 took 0.0434 seconds
10 iterations of einsum 1 took 0.4333 seconds; avg 0.0433 secs
```
